forgerock-openbanking-upgrade/src/main/java/com/forgerock/openbanking/upgrade/UpgradeDatabase.java

```diff
@@ -43,18 +44,26 @@ public CommandLineRunner upgrade() {
     }
 
     @Override
-    public EntitiesVersion getPreviousVersion() {
-        List<EntitiesVersion> all = entitiesVersionRepository.findAll();
+    public EntitiesVersion getStoredVersion() {
+        List<EntitiesVersion> entitiesVersions = cleanDuplicatedVersions(entitiesVersionRepository.findAll());
         EntitiesVersion entitiesVersion;
-        if (all.isEmpty()) {
+        if (entitiesVersions.isEmpty()) {
             entitiesVersion = new EntitiesVersion();
             entitiesVersion.setStatus(EntitiesVersion.UpgradeStatus.UPGRADED);
             entitiesVersion.setVersion("0");
-        } else if (all.size() > 1) {
-            LOGGER.error("More than one version found: " + all);
-            throw new RuntimeException("More than one version found: " + all);
+        } else if (entitiesVersions.size() > 1) {
+            Optional<EntitiesVersion> versionOptional = entitiesVersionRepository.findByVersionAndStatus(
+                    buildProperties.getVersion(),
+                    EntitiesVersion.UpgradeStatus.UPGRADED.name()
+            );
+            if (versionOptional.isPresent()) {
+                LOGGER.info("Current {} version {} found", EntitiesVersion.UpgradeStatus.UPGRADED.name(), buildProperties.getVersion());
+                return versionOptional.get();
```

Comment:
As from the previous comment, is it possible that a cleaned version has the status `UpgradeStatus.UPGRADED` leaving `UpgradeStatus.IN_PROCESS` behind which would throw an error here?

Reply:
This is a double check after clean duplications, probably never happens because in this point the list of versions must have only one version, we don't store the old versions, it means that we don't have a version history in entitiesVersion, the entitiesVersions only must store one entry, the current version that it will be the previous version in the next deployment.
We could improve the current upgrade process but the priority is sbat.

Reply:
Remember there may be two processes running which may both be access this datastore!